demos/sample_azure_iot_pnp/sample_azure_iot_pnp.c

```diff
@@ -0,0 +1,1062 @@
+/* Copyright (c) Microsoft Corporation. All rights reserved. */
+/* SPDX-License-Identifier: MIT */
+
+/* Standard includes. */
+#include <string.h>
+#include <stdio.h>
+
+/* Kernel includes. */
+#include "FreeRTOS.h"
+#include "task.h"
+
+/* Azure Provisioning/IoT Hub library includes */
+#include "azure_iot_hub_client.h"
+#include "azure_iot_provisioning_client.h"
+#include "azure_iot_hub_client_properties.h"
+
+/* Azure JSON includes */
+#include "azure_iot_json_reader.h"
+#include "azure_iot_json_writer.h"
+
+/* Exponential backoff retry include. */
+#include "backoff_algorithm.h"
+
+/* Transport interface implementation include header for TLS. */
+#include "transport_tls_socket.h"
+
+/* Crypto helper header. */
+#include "crypto.h"
+
+/* Demo Specific configs. */
+#include "demo_config.h"
+/*-----------------------------------------------------------*/
+
+/* Compile time error for undefined configs. */
+#if !defined( democonfigHOSTNAME ) && !defined( democonfigENABLE_DPS_SAMPLE )
+    #error "Define the config democonfigHOSTNAME by following the instructions in file demo_config.h."
+#endif
+
+#if !defined( democonfigENDPOINT ) && defined( democonfigENABLE_DPS_SAMPLE )
+    #error "Define the config dps endpoint by following the instructions in file demo_config.h."
+#endif
+
+#ifndef democonfigROOT_CA_PEM
+    #error "Please define Root CA certificate of the IoT Hub(democonfigROOT_CA_PEM) in demo_config.h."
+#endif
+
+#if defined( democonfigDEVICE_SYMMETRIC_KEY ) && defined( democonfigCLIENT_CERTIFICATE_PEM )
+    #error "Please define only one auth democonfigDEVICE_SYMMETRIC_KEY or democonfigCLIENT_CERTIFICATE_PEM in demo_config.h."
+#endif
+
+#if !defined( democonfigDEVICE_SYMMETRIC_KEY ) && !defined( democonfigCLIENT_CERTIFICATE_PEM )
+    #error "Please define one auth democonfigDEVICE_SYMMETRIC_KEY or democonfigCLIENT_CERTIFICATE_PEM in demo_config.h."
+#endif
```

Comment:
not for this PR but we should have demo_config_check.h defined and include in sample